python/cuml/cuml/model_selection/_split.py

```diff
@@ -363,55 +373,63 @@ def train_test_split(
         train_indices = range(0, train_size)
         test_indices = range(-1 * test_size, 0)
 
-    # Gather from indices
-    X_train = X_arr[train_indices]
-    X_test = X_arr[test_indices]
-    if y is not None:
-        y_train = y_arr[train_indices]
-        y_test = y_arr[test_indices]
+    if all_numeric:
+        # Gather from indices
+        X_train = X_arr[train_indices]
+        X_test = X_arr[test_indices]
+        if y is not None:
+            y_train = y_arr[train_indices]
+            y_test = y_arr[test_indices]
+
+        # Coerce output to original input type
+        if ty := determine_df_obj_type(X):
+            x_type = ty
+        else:
+            x_type = determine_array_type(X)
 
-    # Coerce output to original input type
-    if ty := determine_df_obj_type(X):
-        x_type = ty
-    else:
-        x_type = determine_array_type(X)
+        if ty := determine_df_obj_type(y):
+            y_type = ty
+        else:
+            y_type = determine_array_type(y)
 
-    if ty := determine_df_obj_type(y):
-        y_type = ty
-    else:
-        y_type = determine_array_type(y)
+        if x_type in ("series", "dataframe"):
+            X_train = output_to_df_obj_like(X_train, X, x_type)
+            X_test = output_to_df_obj_like(X_test, X, x_type)
+
+            if determine_array_type(X.index) == "pandas":
+                if isinstance(train_indices, cp.ndarray):
+                    train_indices = train_indices.get()
+                if isinstance(test_indices, cp.ndarray):
+                    test_indices = test_indices.get()
+
+            X_train.index = X.index[train_indices]
+            X_test.index = X.index[test_indices]
+        else:
+            X_train = X_train.to_output(x_type)
+            X_test = X_test.to_output(x_type)
+
+        if y_type in ("series", "dataframe"):
+            y_train = output_to_df_obj_like(y_train, y, y_type)
+            y_test = output_to_df_obj_like(y_test, y, y_type)
 
-    if x_type in ("series", "dataframe"):
-        X_train = output_to_df_obj_like(X_train, X, x_type)
-        X_test = output_to_df_obj_like(X_test, X, x_type)
+            if determine_array_type(y.index) == "pandas":
+                if isinstance(train_indices, cp.ndarray):
```

Comment:
We ought to deduplicate this logic. It is not necessary to write the same steps twice for X and y.